design/one-pager-composition-environment.md

```diff
@@ -0,0 +1,266 @@
+* `provider-kubernetes`: using [`spec.forProvider.references[].patchesFrom`](https://doc.crds.dev/github.com/crossplane-contrib/provider-kubernetes/kubernetes.crossplane.io/Object/v1alpha1@v0.3.0#spec-references-patchesFrom)
+
+However, none of these methods is native to Crossplane native and they are not
+available for every provider. Furthermore, they would require an additional
+managed resource (here `Object` or `Release`) to deploy the composed resource
+making the composition much more complex.
+
+## Goals
+
+The goal of this document is to define an API design for extracting environment
+specific data considering the following requirements:
+
+1. Allow the creation of generic compositions that render composed resources
+based on the environment they are executed in.
+2. Allows users to define this environment in Crossplane native way.
+3. Prevent escalation of priveleges by disallowing users to read out fields
+from objects they shouldn't have access to.
+
+## Proposed Solution
+
+Introduce a new Crossplane resource `EnvironmentConfig` that can store any 
+kind of  data similar to a K8s `ConfigMap` but supports complex data as well
+using the [`JSON` struct](https://pkg.go.dev/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1#JSON)
+which can store anything that serializes to valid JSON.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: EnvironmentConfig
+metadata:
+  name: example-environment
+data:
+  simple: value
+  int: 123
+  bool: false
+  complex:
+    a: b
+    c:
+      d: e
+  list:
+    - a
+    - b
+    - c
+```
+
+`EnvironmentConfig`s are referenced in the composition spec. They can be
+referenced directly as names or through labels in `spec.environment.environmentConfigs`.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: example
+spec:
+  # This feature will be introduced as alpha (behind a flag). Composition is a
+  # v1 API so we'll need to add a note that the environment block is alpha and
+  # may change without notice.
+  environment:
+    environmentConfigs:
+    # The EnvironmentConfigs this Composition will use. The data of these
+    # EnvironmentConfigs is merged into a single 'computed' environment object,
+    # that is written to status.environment.computed. This computed environment
+    # is what all environment patches operate upon. 
+    - type: Reference
+      reference:
+        name: example-environment
+    - type: Reference
+      reference:
+        name: other-environment
+    # Each selector in this array matches at most one EnvironmentConfig. If a
+    # selector matches multiple we pick the first. (This is how cross resource
+    # ref selectors work.) Each matching EnvironmentConfig is 'pinned' as a
+    # reference below. Selecting is optional; you can just reference one or more
```

Comment:
I think we now pin these to the list of references on an XR, not "below" on the Composition.

Reply:
That's corrent. I removed this section.